Runtime hooks: Fixes #2759, multiprocess spawn mode on POSIX OSs

bpteague opened the pull request:
This PR is based on @springermac's #2793. It fixes issues #2322 and #2759. It supersedes #2505.

See my comments in #2322 for a diagnosis of the underlying issue.

This PR adds a runtime hook that patches `multiprocessing.freeze_support` to catch the Python library's attempts to start the semaphore tracker and fork server processes.  It also includes the logic from https://github.com/pyinstaller/pyinstaller/wiki/Recipe-Multiprocessing.

springermac commented on tests/functional/test_multiprocess.py:
```diff
@@ -0,0 +1,173 @@
+
+
+@skipif(not is_py34, reason='set_start_method introduced in python 3.4+')
+@skipif(sys.platform.startswith('win'), reason='fork is not available on windows')
```
Please use the convenience methods from PyInstaller.utils.tests or PyInstaller.compat.

bpteague replied:
Got it. Thanks!

springermac commented on tests/functional/test_multiprocess.py:
```diff
@@ -0,0 +1,173 @@
+
+
+@skipif(not is_py34, reason='set_start_method introduced in python 3.4+')
+@skipif(sys.platform.startswith('win'), reason='forkserver is not available on windows')
```
Again convenience methods from utils.tests or compat.

bpteague replied:
Got it.  Thanks!

bpteague commented:
@springermac Done

htgoebel commented on PyInstaller/loader/rthooks/pyi_rth_multiprocessing.py:
```diff
@@ -1,22 +1,23 @@
+                set(sys.argv[1:-2]) == set(_args_from_interpreter_flags()) and \
+                sys.argv[-2] == '-c' and \
+                ('multiprocessing.semaphore_tracker' in sys.argv[-1] or \
+                 'multiprocessing.forkserver' in sys.argv[-1]):
```
- This looks like magic. Please add a comment about that this code does.
- `'multiprocessing.forkserver' in sys.argv[-1]` would allow 'xxxxxmultiprocessing.forkserveryyyyyy'. Is this intended?

bpteague replied:
I've added an extensive comment, and changed the string matching to more closely match what comes out of the Python standard library.  Unfortunately, this is quite closely linked to the private behavior of the multiprocessing module :-/  (Really, the native freeze_support should be doing this.  I'll submit a bug to upstream in a bit.)

bpteague commented on PyInstaller/loader/rthooks/pyi_rth_multiprocessing.py:
```diff
@@ -1,8 +1,7 @@
 # 'spawn' multiprocessing needs some adjustments on osx
-if _platform == 'darwin':
-    import sys
+if (sys.platform == 'darwin' or sys.platform == 'linux') and sys.version_info >= (3, 4):
```
This line is removed in a later commit; we should monkey-patch on all platforms.

bpteague commented:
@htgoebel - thanks for your feedback!  I've added an extensive comment explaining the "magic" and rebased as you requested.  Let me know if there's anything else you'd like.

springermac commented on PyInstaller/loader/rthooks/pyi_rth_multiprocessing.py:
```diff
@@ -1,22 +1,35 @@
+                set(sys.argv[1:-2]) == set(_args_from_interpreter_flags()) and \
+                sys.argv[-2] == '-c' and \
+                (sys.argv[-1].startswith('from multiprocessing.semaphore_tracker import main') or \
+                 sys.argv[-1].startswith('from multiprocessing.forkserver import main'):
```
Syntax error. Missing parenthesis.

bpteague commented:
@htgoebel or @springermac could you kick Travis?

springermac commented:
@bpteague Close and reopen the pull request. It will restart all the tests.

htgoebel commented:
Thank you very much for the updated commits :-)

christianbioinf commented:
@htgoebel I have an issue that seems to be related to this hook using pyinstaller 3.3.1 with python 3.4.3. On OSX and on linux, I have a frozen one-file bundle (i.e. main) that executes frozen one-file bundles (i.e. subs) via multiprocessing. When I use fork as start_method it does not work since the subs have the _MEIPASS2 environment variable already set and hence do not extract themselves and fail with library loading errors. Everything works fine when using the spawn method (and the freeze_support) or when I do not use multiprocessing but call subprocess.Popen in the main directly (unfortunately there is no multiprocessing then). Could I undo the hook somehow for my case or why do you need to set the _MEIPASS2 environment variable anyways in case of linux/osx since from the history of this issue it seemed to be releated to windows only.

htgoebel commented:
@charlesnicholson If you have an issue, please open a ticket. This pull-request is closed since six month already.

charlesnicholson commented:
I think you meant to tag @christianbioinf, I'm unfamiliar with this issue...

htgoebel commented:
@charlesnicholson Oh soory.

christianbioinf commented:
@htgoebel Thanks for the advice. I have now opened a new issue #3409 including a gist to easily reproduce the issue.
